lib/xnvme_be_macos_driverkit_mem.c

```diff
@@ -0,0 +1,93 @@
+// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <libxnvme.h>
+#include <xnvme_be.h>
+#include <xnvme_be_nosys.h>
+#ifdef XNVME_BE_MACOS_ENABLED
+#include <xnvme_be_macos_driverkit.h>
+
+#include <mach/mach_error.h>
+
+#include <xnvme_dev.h>
+#include <errno.h>
+#include <sys/mman.h>
+
+void *
+xnvme_be_macos_driverkit_buf_alloc(const struct xnvme_dev *dev, size_t nbytes,
+				   uint64_t *XNVME_UNUSED(phys))
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+	long sz = sysconf(_SC_PAGESIZE);
+
+	if (sz == -1) {
+		XNVME_DEBUG("FAILED: sysconf(), errno: %d", errno);
+		return NULL;
+	}
+	XNVME_DEBUG("Page size: %lx", sz);
+
+	// nbytes has to be a multiple of alignment. Therefore, we round up to the nearest
+	// multiple.
+	XNVME_DEBUG("Unaligned size: %lx", nbytes);
+	nbytes = (1 + ((nbytes - 1) / sz)) * (sz);
+	XNVME_DEBUG("Aligned size: %lx", nbytes);
+
+	void *ptr = xnvme_buf_virt_alloc(sz, nbytes);
+
+	uint64_t *buf = (uint64_t *)ptr;
+	buf[0] = nbytes;        // Put actual size inside allocated memory. Driver on the other end
+				// doesn't correctly get this.
+	buf[1] = (uint64_t)ptr; // Our token for the memory reference
+
+	XNVME_DEBUG("buf token: %llx", (uint64_t)ptr);
+	kern_return_t ret = IOConnectCallStructMethod(state->device_connection, NVME_ALLOC_BUFFER,
+						      ptr, nbytes, NULL, 0);
+	assert(ret == kIOReturnSuccess);
+
+	buffer_add(state->buffers, (struct buffer){.vaddr = (uint64_t)ptr, .nbytes = nbytes});
+	return ptr;
+}
+
+void
+xnvme_be_macos_driverkit_buf_free(const struct xnvme_dev *dev, void *buf)
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+	if (!buf) {
+		return;
+	}
+	struct buffer *_buf = buffer_find(state->buffers, (uint64_t)buf);
+	assert(_buf);
+
+	XNVME_DEBUG("buf token: %llx", (uint64_t)_buf->vaddr);
```

Comment:
See notes on declaration / initialization order and error-handling.

Reply:
Done.